### PR TITLE
Add initial jujud operator

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

Add a skeleton jujud operator. The operator starts essentially a no-op worker which can be fleshed out going forward. Some small common bits from uniter were extracted out, namely the symlink creation.

The CAAS provisioner facade and state model gets support for setting a password on the application entity. This will allow the operator via the associated facade (not yet written) to make api calls back to the controller.

The k8s client is improved to delete and re-create config map and pod so allow for changed config to be applied as needed. The pod delete gets a retry loop to allow for the terminating phase to complete.

## QA steps

Run up a CAAS model and deploy a charm.
Run upgrade-juju and check that the operator pod is deleted and restarted.

### CODE REVIEW

**Comment by axw** [comment]
Please split this up into multiple commits, at least the following:
 - state changes
 - uniter refactoring
 - worker/caasoperator
 - cmd/jujud

**Comment by wallyworld** [comment]
!!build!!

File: state/migration_internal_test.go [hunk 1]
```diff
@@ -368,6 +368,8 @@ func (s *MigrationSuite) TestApplicationDocFields(c *gc.C) {
 		// RelationCount is handled by the number of times the application name
 		// appears in relation endpoints.
 		"RelationCount",
+		// TODO(caas) - add to export/import
```
Comment (axw) [inline state/migration_internal_test.go]: please create a bug and/or card so it doesn't get forgotten

File: api/caasprovisioner/client.go [hunk 2]
```diff
@@ -8,6 +8,7 @@ import (
 	apiwatcher "github.com/juju/juju/api/watcher"
 	"github.com/juju/juju/apiserver/params"
 	"github.com/juju/juju/watcher"
+	"gopkg.in/juju/names.v2"
```
Comment (axw) [inline api/caasprovisioner/client.go]: move me

File: api/caasprovisioner/client.go [hunk 3]
```diff
@@ -25,14 +26,37 @@ func NewClient(caller base.APICaller) *Client {
...
+	for i, p := range appPasswords {
+		args.Changes[i] = params.EntityPassword{
+			Tag: names.NewApplicationTag(p.Name).String(), Password: p.Password,
+		}
+	}
+	err := c.facade.FacadeCall("SetPasswords", args, &result)
+	if err != nil {
+		return err
+	}
+	return result.Combine()
```
Comment (axw) [inline api/caasprovisioner/client.go]: each change is independent, so the errors shouldn't be combined
also, check the number of results is the same as the number of args?

File: apiserver/facades/agent/caasprovisioner/provisioner.go [hunk 4]
```diff
@@ -34,11 +38,17 @@ func NewCAASProvisionerAPI(
 	if !authorizer.AuthController() {
 		return nil, common.ErrPerm
 	}
+	getAuthFunc := func() (common.AuthFunc, error) {
```
Comment (axw) [inline apiserver/facades/agent/caasprovisioner/provisioner.go]: the whole API requires controller, so maybe just pass AuthAlways into NewPasswordChanger?

File: apiserver/facades/agent/caasprovisioner/provisioner_test.go [hunk 5]
```diff
@@ -64,3 +65,25 @@ func (s *CAASProvisionerSuite) TestWatchApplications(c *gc.C) {
 	c.Assert(resource, gc.NotNil)
 	c.Assert(resource, gc.Implements, new(state.StringsWatcher))
 }
+
+func (s *CAASProvisionerSuite) TestSetPasswords(c *gc.C) {
+	s.st.app = &mockApplication{
+		tag: names.NewApplicationTag("app")}
+	s.authorizer.Controller = true
```
Comment (axw) [inline apiserver/facades/agent/caasprovisioner/provisioner_test.go]: already set in SetUpTest

File: apiserver/facades/agent/caasprovisioner/provisioner_test.go [hunk 6]
```diff
@@ -64,3 +65,25 @@ func (s *CAASProvisionerSuite) TestWatchApplications(c *gc.C) {
 	c.Assert(resource, gc.NotNil)
 	c.Assert(resource, gc.Implements, new(state.StringsWatcher))
 }
+
+func (s *CAASProvisionerSuite) TestSetPasswords(c *gc.C) {
+	s.st.app = &mockApplication{
+		tag: names.NewApplicationTag("app")}
```
Comment (axw) [inline apiserver/facades/agent/caasprovisioner/provisioner_test.go]: ,\n}

File: caas/kubernetes/provider/k8s.go [hunk 7]
```diff
@@ -84,23 +90,25 @@ func (k *kubernetesClient) EnsureOperator(appName, agentPath string, newConfig c
...
-	} else {
-		config, err := newConfig(appName)
-		if err != nil {
-			return "", errors.Annotate(err, "creating config")
-		}
-		if err := k.createConfigMap(mapName, config); err != nil {
-			return "", errors.Annotate(err, "creating ConfigMap")
+	} else if exists {
+		logger.Debugf("ConfigMap %s already exists - deleting", mapName)
+		if err := k.deleteConfigMap(mapName); err != nil {
```
Comment (axw) [inline caas/kubernetes/provider/k8s.go]: can we just update the config map if it already exists? rather than making 2 API calls

File: cmd/jujud/agent/caasoperator_test.go [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+type CAASOperatorSuite struct {
+	coretesting.GitSuite
+	agenttest.AgentSuite
+}
+
+var _ = gc.Suite(&CAASOperatorSuite{})
+
+func (s *CAASOperatorSuite) SetUpSuite(c *gc.C) {
+	s.GitSuite.SetUpSuite(c)
```
Comment (axw) [inline cmd/jujud/agent/caasoperator_test.go]: Is this GitSuite needed? I'm guessing you cribbed this from unit_test.go. I don't think the CAAS operator or tests should need it.

File: cmd/jujud/agent/caasoperator_test.go [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+	s.AgentSuite.SetUpSuite(c)
```
Comment (axw) [inline cmd/jujud/agent/caasoperator_test.go]: Please don't copy the existing agent tests, they're awful. Can you please make these real unit tests, without an embedded JujuConnSuite? We can add functional tests a bit later, in the feature tests package.

File: cmd/jujud/agent/caasoperator_test.go [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+	}
+
+	err = a.Init(nil)
+	c.Assert(err, gc.IsNil)
+
+	_, ok := ctx.Stderr.(*lumberjack.Logger)
+	c.Assert(ok, jc.IsFalse)
+}
+
+func (s *CAASOperatorSuite) TestWorkers(c *gc.C) {
```
Comment (axw) [inline cmd/jujud/agent/caasoperator_test.go]: I think we should replace this with a combination of manifold name tests (i.e. in the manifold package), and feature tests.

File: cmd/jujud/agent/engine_test.go [hunk 11]
```diff
@@ -133,6 +134,15 @@ var (
 		"unconverted-api-workers",
 		"unit-agent-deployer",
 	}
+
+	alwaysCAASWorkers = []string{
+		"agent",
+		"api-caller",
+		"operator",
```
Comment (axw) [inline cmd/jujud/agent/engine_test.go]: surely "operator" shouldn't be running during migration? or is the TODO because we don't handle migration in the CAAS operator yet?

File: worker/caasoperator/caasoperator.go [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+// A CaasOperatorExecutionObserver gets the appropriate methods called when a hook
+// is executed and either succeeds or fails.  Missing hooks don't get reported
+// in this way.
+type CaasOperatorExecutionObserver interface {
+	HookCompleted(hookName string)
+	HookFailed(hookName string)
+}
+
+// CaasOperator implements the capabilities of the caasoperator agent. It is not intended to
```
Comment (axw) [inline worker/caasoperator/caasoperator.go]: None of this is tested, and it doesn't do anything useful yet anyway. Can you please pull this worker into another PR, with tests for the bits that are there? It doesn't need to be useful, but adding tests as we go means we don't end up piling on top of potentially shaky foundations.

Comment (wallyworld) [reply worker/caasoperator/caasoperator.go]: It's tested in the jujud agent tests where we check that the symlinks are created. So there is test coverage for the bit that is implemented. We will add unit tests as the worker is fleshed out to do more than just start the loop and wait for destruction. The idea of this PR is to add the foundation on which to build up new stuff.

File: cmd/jujud/agent/caasoperator_test.go (hunk 8 above)
Comment (wallyworld) [reply cmd/jujud/agent/caasoperator_test.go]: yeah, not needed

File: apiserver/facades/agent/caasprovisioner/provisioner_test.go (hunk 5 above)
Comment (wallyworld) [reply apiserver/facades/agent/caasprovisioner/provisioner_test.go]: yup

File: apiserver/facades/agent/caasprovisioner/provisioner_test.go (hunk 6 above)
Comment (wallyworld) [reply apiserver/facades/agent/caasprovisioner/provisioner_test.go]: done

File: apiserver/facades/agent/caasprovisioner/provisioner.go (hunk 4 above)
Comment (wallyworld) [reply apiserver/facades/agent/caasprovisioner/provisioner.go]: done

File: api/caasprovisioner/client.go (hunk 3 above)
Comment (wallyworld) [reply api/caasprovisioner/client.go]: fixed

File: api/caasprovisioner/client.go (hunk 2 above)
Comment (wallyworld) [reply api/caasprovisioner/client.go]: done

File: state/migration_internal_test.go (hunk 1 above)
Comment (wallyworld) [reply state/migration_internal_test.go]: created card

File: cmd/jujud/agent/engine_test.go (hunk 11 above)
Comment (wallyworld) [reply cmd/jujud/agent/engine_test.go]: correct, there's no migration checks for CAAS models yet

File: caas/kubernetes/provider/k8s.go (hunk 7 above)
Comment (wallyworld) [reply caas/kubernetes/provider/k8s.go]: yep

File: cmd/jujud/agent/caasoperator_test.go (hunk 10 above)
Comment (wallyworld) [reply cmd/jujud/agent/caasoperator_test.go]: The unit tests have been left in place and this moved to a feature test. It tests the machinery that starts the workers so best to retain it (as well as existing name tests).

File: cmd/jujud/agent/caasoperator_test.go (hunk 9 above)
Comment (wallyworld) [reply cmd/jujud/agent/caasoperator_test.go]: Took out JujuConnSuite, left in tests which are unit tests

File: cmd/jujud/agent/caasoperator_test.go [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+func (s *CAASOperatorSuite) newBufferedLogWriter() *logsender.BufferedLogWriter {
+	logger := logsender.NewBufferedLogWriter(1024)
+	s.AddCleanup(func(*gc.C) { logger.Close() })
+	return logger
+}
+
+func (s *CAASOperatorSuite) TestParseSuccess(c *gc.C) {
+	//	s.primeAgent(c)
```
Comment (axw) [inline cmd/jujud/agent/caasoperator_test.go]: delete?

File: cmd/jujud/agent/agenttest/engine.go [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+	goyaml "gopkg.in/yaml.v2"
+
+	coretesting "github.com/juju/juju/testing"
+	"github.com/juju/juju/worker/dependency"
+)
+
+// NewWorkerManager takes an EngineTracker, an engine manager id to
+// monitor and the workers that are expected to be running and sets up
+// a WorkerManager.
+func NewWorkerMatcher(c *gc.C, tracker *EngineTracker, id string, workers []string) *WorkerMatcher {
```
[the remark this answers is not in the document]
Comment (axw) [reply cmd/jujud/agent/agenttest/engine.go]: actually I think maybe the comment is just wrong?

[the remark this answers is not in the document]
Comment (wallyworld) [reply cmd/jujud/agent/agenttest/engine.go]: yeah, comment is wrong

File: cmd/jujud/agent/caasoperator.go [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+	defer op.tomb.Done()
+	if err := op.ReadConfig(op.Tag().String()); err != nil {
+		return err
+	}
+	logger.Infof("caas operator %v start (%s [%s])", op.Tag().String(), jujuversion.Current, runtime.Compiler)
+	if flags := featureflag.String(); flags != "" {
+		logger.Warningf("developer feature flags enabled: %s", flags)
+	}
+
+	op.runner.StartWorker("api", op.APIWorkers)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply cmd/jujud/agent/caasoperator.go]: agreed

File: cmd/jujud/agent/caasoperator_test.go (hunk 13 above)
Comment (wallyworld) [reply cmd/jujud/agent/caasoperator_test.go]: done

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$
